### PR TITLE
Spring boot practice category

### PR DESCRIPTION
**Pull request by SerhiiAN**

### CODE REVIEW

File: src/main/java/mate/academy/intro/controller/BookController.java [hunk 1]
```diff
@@ -34,6 +36,7 @@
 public class BookController {
     private final BookService bookService;
 
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/controller/BookController.java]: replace to service layer

File: src/main/java/mate/academy/intro/dto/book/CreateBookRequestDto.java [hunk 2]
```diff
@@ -26,6 +27,6 @@ public class CreateBookRequestDto {
     @NotBlank(message = "The cover image is mandatory.")
     @Pattern(regexp = "^(https?:\\/\\/.*\\.(?:png|jpg|jpeg|gif))$",
             message = "not the correct format for cover image")
-
     private String coverImage;
+    private Set<Long> categoryIds;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/dto/book/CreateBookRequestDto.java:33]: add validation

File: src/main/java/mate/academy/intro/model/Book.java [hunk 3]
```diff
@@ -35,4 +43,12 @@ public class Book {
     @Column(nullable = false,
             columnDefinition = "TINYINT(1) DEFAULT 0")
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(
+            name = "book_category",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/model/Book.java]:
```suggestion
            name = "books_categories",

```

File: src/main/java/mate/academy/intro/model/Category.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.persistence.ManyToMany;
+import jakarta.persistence.Table;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Table(name = "category")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/model/Category.java]:
```suggestion
@Table(name = "categories")
```

File: src/main/java/mate/academy/intro/model/Category.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @Column(nullable = false,
+            columnDefinition = "TINYINT(1) DEFAULT 0")
+    private boolean isDeleted = false;
+    @ManyToMany(mappedBy = "categories", fetch = FetchType.LAZY)
+    private Set<Book> books = new HashSet<>();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/model/Category.java]:
```suggestion
```

File: src/main/java/mate/academy/intro/model/Category.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @Column(nullable = false,
+            columnDefinition = "TINYINT(1) DEFAULT 0")
+    private boolean isDeleted = false;
+    @ManyToMany(mappedBy = "categories", fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/model/Category.java]:
```suggestion
```

File: src/main/java/mate/academy/intro/service/category/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable)
+                .stream()
+                .map(categoryMapper::toDto).toList();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/category/CategoryServiceImpl.java]:
```suggestion
                .map(categoryMapper::toDto)
                .toList();
```

File: src/main/resources/db/changelog/changes/08-create-category-table-and-change-book-table.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-category-table
+      author: serhii
+      changes:
+        - createTable:
+            tableName: category
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-create-category-table-and-change-book-table.yaml]:
```suggestion
            tableName: categories
```

File: src/main/resources/db/changelog/changes/08-create-category-table-and-change-book-table.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
+  - changeSet:
+      id: create-book-category-join-table
+      author: serhii
+      changes:
+        - createTable:
+            tableName: book_category
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-create-category-table-and-change-book-table.yaml]:
```suggestion
            tableName: books_categories
```

File: src/main/resources/db/changelog/changes/08-create-category-table-and-change-book-table.yaml [hunk 10]
```diff
@@ -0,0 +1,59 @@
+databaseChangeLog:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-create-category-table-and-change-book-table.yaml]: rename file, you don't change book table
it's better create separate changelogs for create category and create book-category

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/mate/academy/intro/service/BookServiceImpl.java [hunk 11]
```diff
@@ -27,12 +29,14 @@ public BookDto save(CreateBookRequestDto requestDto) {
         return bookMapper.toDto(bookRepository.save(book));
     }
 
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/intro/service/BookServiceImpl.java [hunk 12]
```diff
@@ -27,12 +29,14 @@ public BookDto save(CreateBookRequestDto requestDto) {
...
     }
 
+    @Transactional
     @Override
     public List<BookDto> findAll(Pageable pageable) {
         return bookRepository.findAll().stream()
                 .map(bookMapper::toDto).toList();
     }
 
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/intro/service/BookServiceImpl.java [hunk 13]
```diff
@@ -55,11 +59,21 @@ public void deleteById(Long id) {
...
     public List<BookDto> search(BookSearchParameters parameters, Pageable pageable) {
         Specification<Book> bookSpecification = bookSpecificationBuilder.build(parameters);
         return bookRepository.findAll(bookSpecification, pageable)
                 .stream()
                 .map(bookMapper::toDto)
                 .toList();
     }
+
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/intro/service/BookServiceImpl.java [hunk 14]
```diff
@@ -55,11 +59,21 @@ public void deleteById(Long id) {
         bookRepository.deleteById(id);
     }
 
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/intro/service/BookServiceImpl.java (hunk 11 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/intro/service/BookServiceImpl.java]: Add `@Transactional `for save method, also add there
`book.setCategories.....`

Add `@Transactional` for update method

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/intro/dto/book/CreateBookRequestDto.java [hunk 15]
```diff
@@ -26,6 +27,7 @@ public class CreateBookRequestDto {
     @NotBlank(message = "The cover image is mandatory.")
     @Pattern(regexp = "^(https?:\\/\\/.*\\.(?:png|jpg|jpeg|gif))$",
             message = "not the correct format for cover image")
-
     private String coverImage;
+    @NotNull(message = "At least one category ID is required.")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/dto/book/CreateBookRequestDto.java]:
```suggestion
    @NotEmpty(message = "At least one category ID is required.")
```

File: src/main/java/mate/academy/intro/controller/BookController.java [hunk 16]
```diff
@@ -78,7 +79,7 @@ public void deleteById(@PathVariable @Positive Long id) {
     @PreAuthorize("hasRole('USER')")
     @GetMapping("/search")
     @Operation(summary = "Search books", description = "Search for books based on various criteria")
-    public List<BookDto> search(BookSearchParameters searchParameters,
+    public List<BookDto> search(@ModelAttribute BookSearchParameters searchParameters,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/controller/BookController.java]: Why do you need `@ModelAttribute`?

File: src/main/java/mate/academy/intro/model/Book.java [hunk 17]
```diff
@@ -35,4 +43,12 @@ public class Book {
     @Column(nullable = false,
             columnDefinition = "TINYINT(1) DEFAULT 0")
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/model/Book.java]:
```suggestion
    @ManyToMany
```

Lazy by default

File: src/main/java/mate/academy/intro/service/BookServiceImpl.java [hunk 18]
```diff
@@ -21,12 +23,14 @@ public class BookServiceImpl implements BookService {
...
     private final BookSpecificationBuilder bookSpecificationBuilder;
 
+    @Transactional
     @Override
     public BookDto save(CreateBookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
         return bookMapper.toDto(bookRepository.save(book));
     }
 
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/BookServiceImpl.java]: remove

File: src/main/java/mate/academy/intro/service/BookServiceImpl.java [hunk 19]
```diff
@@ -62,4 +67,12 @@ public List<BookDto> search(BookSearchParameters parameters, Pageable pageable)
                 .map(bookMapper::toDto)
                 .toList();
     }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/BookServiceImpl.java:69]: Why method search is absent in BookService interface?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

File: src/main/java/mate/academy/intro/repository/category/CategoryRepository.java [hunk 20]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.intro.repository.category;
+
+import mate.academy.intro.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/intro/repository/category/CategoryRepository.java]:
```suggestion

```

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/intro/repository/book/BookRepository.java [hunk 21]
```diff
@@ -1,10 +1,15 @@
...
 
+import java.util.List;
 import mate.academy.intro.model.Book;
+import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/intro/repository/book/BookRepository.java]:
```suggestion

```

Comment (MentorMateAcademy) [reply src/main/java/mate/academy/intro/repository/book/BookRepository.java]: you can clear this anno

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/intro/controller/CategoryController.java [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category management", description = "Endpoints for managing categories")
+@RequiredArgsConstructor
+@Validated
+@RestController
+@RequestMapping("/api/categories")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/controller/CategoryController.java]:
```suggestion
@RequestMapping("/categories")

```

File: src/main/java/mate/academy/intro/model/Book.java [hunk 23]
```diff
@@ -35,4 +42,12 @@ public class Book {
     @Column(nullable = false,
             columnDefinition = "TINYINT(1) DEFAULT 0")
     private boolean isDeleted = false;
+    @ManyToMany
+    @JoinTable(
+            name = "book_categories",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/model/Book.java]:
```suggestion
            name = "books_categories",
```

File: src/main/java/mate/academy/intro/model/Category.java [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Table(name = "categories")
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE category SET is_deleted = TRUE WHERE id=?")
+@Where(clause = "is_deleted=FALSE")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/model/Category.java]: `@Where` is deprecated

File: src/main/java/mate/academy/intro/service/user/UserServiceImpl.java [hunk 25]
```diff
@@ -27,6 +27,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         }
         User user = userMapper.toModel(requestDto);
         user.setPassword(passwordEncoder.encode(user.getPassword()));
-        return userMapper.toUserResponseDto(userRepository.save(user));
+        UserResponseDto responseDto = userMapper.toUserResponseDto(userRepository.save(user));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/intro/service/user/UserServiceImpl.java]: why do you need to make this changes?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by okuzan — APPROVED** [review]

**Pull request merged**